### PR TITLE
[Triton/Gluon] [ASM] [HIP] Mha v4 fixes sparse etc

### PR DESCRIPTION
**Pull request by jcaraban**
⚠️ These MHA kernels are mainly tested & intended for Diffusion Inference workloads.
However, mha_v4 gfx950 bf16 is virtually as accurate as mha_v3 at +100 TFlops 👌

## Motivation

Add log-sum-exp output to MHA v4 gfx950 kernels (only dense variants now), so it can run under ring / context parallelism, which merges per-rank partials via LSE. Also extends block-sparse to the BF16 recipes, canonicalizes the MXFP4 rows, and improves K/V quantization accuracy.

## Technical Details

Kernels

- New BF16 and BF16FP8 sorted-sparse kernels.
- Dense LSE epilogue on all ten hd128 recipes, gated at runtime on `s_lse`.
- `f8f6`, `f6f4` and `mxfp4` sparse rows moved to FP6-P V, matching their dense siblings.
- Dense `mxfp4` claims the canonical FP6-P V order; the duplicate `f4f4` row is disabled.

Host

- bugfix: per-channel V amax clamped so empty heads cannot quantize to NaN.
- Optional `lse` output on `mha_v4` / `mha_v4_packed`, plumbed through `asm_mha_v4_fwd.cu`.
  + ABI unchanged: `ptr_lse` / `s_lse` / `s_lse_Hs` were already reserved in the kernarg.
- `_LSE_CAPABLE_QV` gates the supported format pairs; sorted-sparse still raises.
- K mean is subtracted before quantizing (K-smoothing), fused into the MX quantizer kernels.
- Dense MHA v4 accepts per-batch key lengths (ragged `seqlen_k`).

Minor

- Retired `MXFP4 Q/K + FP8 V` and the deprecated `mha_v4_mxfp8` alias.
- `bench_sage.py`: improve input distributions, diffusion-calibrated default, BF16 sparse modes.
- Split the block-sparse cases into `op_tests/test_mha_v4_sparse.py`.

## Test Plan

1. Run the complete MHA v4 test suite on MI355X/gfx950.
2. Per-recipe dense and sparse validation against Torch; full-LUT sparse bit-exact vs dense.
3. Compare LSE against float32 `torch.logsumexp` for every capable format pair.
4. Compare O bitwise, LSE off vs on and against the previous code objects.
5. Interleaved A/B throughput, 12 rounds per variant on one idle GPU.

## Test Result

1. 316 passed, 1 skipped in test_mha_v4.py + test_mha_v4_sparse.py on MI355X/gfx950.
2. Dense validation passes 10/10 recipes; unchanged sparse objects are byte-identical.
3. LSE vs float32: bf16 max_abs 7.2e-4; quantized rows |mean_off| <= 0.046.
 a. The residual is an additive approximate-exp2 bias and cancels in ring combine weights.
4. O bitwise unchanged vs the previous objects (40/40 cases); dense throughput at parity
5. black --check and ruff on PR-touched Python files: passed.

Previous PRs: #5335, #5005, #4967, #4627 

image

### CODE REVIEW

**Review by azaidy — CHANGES_REQUESTED** [review]
LGTM from the triton side overall.
Please reduce the comments in `op_tests/op_benchmarks/triton/bench_sage.py`. They should be 2-3 lines maximum.

File: csrc/py_itfs_cu/asm_mha_v4_fwd.cu [hunk 1]
```diff
@@ -859,6 +870,47 @@ void fmha_v4_fwd(const at::Tensor& q,
...
+                    " for batch ",
+                    shapes.batch);
+        args.ptr_kseq.value = seqlens_k->data_ptr();
+    }
+
+    // LSE is opt-in the same way: the kernels branch on s_lse and skip the store when it is zero,
+    // so a launch without it leaves the reserved slots at zero and keeps the same code object.
+    if(lse.has_value())
+    {
+        TORCH_CHECK(lse->is_cuda(), "MHA v4 lse must be a GPU tensor");
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply csrc/py_itfs_cu/asm_mha_v4_fwd.cu]: addressed in [009c3504f](https://github.com/ROCm/aiter/pull/5798/changes/009c3504fa200479f14e375cf305f33491a16e9f)

File: aiter/ops/mha_v4.md [hunk 2]
```diff
@@ -8,23 +8,28 @@ dispatch metadata. Unsupported recipes fail instead of falling back to another a
...
-| INT8 | FP8 |
-| MXFP8 | FP8 |
-| FP8 | FP8 |
-| FP8 | MXFP6 |
-| MXFP6 | FP8 |
-| MXFP6 | MXFP6 (dense only) |
-| MXFP6 | MXFP4 |
-| MXFP4 | MXFP4 |
+- Per-batch key lengths on dense via `seqlens_k`; sorted sparse rejects it.
+- No backward, dropout, RNG state, LSE, causal, or Q-side varlen support yet.
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply aiter/ops/mha_v4.md]: https://github.com/ROCm/aiter/pull/5798/changes/009c3504fa200479f14e375cf305f33491a16e9f

File: csrc/kernels/mha_v4_quant.cu [hunk 3]
```diff
@@ -771,9 +827,32 @@ void launch_quant(aiter_tensor_t& out,
...
+// and leaves the outputs untouched when the tensor is empty.
+static const float* check_k_mean(const aiter_tensor_t& mean,
+                                 const aiter_tensor_t& input,
+                                 int32_t& heads,
+                                 int32_t& seq_heads)
+{
+    if(mean.numel() == 0)
+        return nullptr;
+    AITER_CHECK(mean.dtype() == AITER_DTYPE_fp32, "K mean must be fp32");
+    AITER_CHECK(mean.is_contiguous(), "K mean must be contiguous");
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply csrc/kernels/mha_v4_quant.cu]: https://github.com/ROCm/aiter/pull/5798/changes/e6d53d5251b721b50ad4715b9d1fceb8d79cdb3e

File: aiter/ops/mha_v4.py [hunk 4]
```diff
@@ -832,6 +871,19 @@ def mha_v4_packed(
...
+                "sorted-sparse MHA v4 does not accept per-batch key lengths yet"
+            )
+        if seqlens_k.dtype != torch.int32 or seqlens_k.device != q.device:
+            raise ValueError(
+                "seqlens_k must be an int32 tensor on the same device as Q"
+            )
+        if seqlens_k.numel() < batch:
+            raise ValueError("seqlens_k needs one entry per batch")
+        if not seqlens_k.is_contiguous():
+            raise ValueError("seqlens_k must be contiguous")
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply aiter/ops/mha_v4.py]: https://github.com/ROCm/aiter/pull/5798/changes/cb9a297dfc162bbebc52a20816654f0f9131d643, this was fixed in kernel

File: csrc/py_itfs_cu/asm_mha_v4_fwd.cu [hunk 5]
```diff
@@ -859,6 +870,47 @@ void fmha_v4_fwd(const at::Tensor& q,
...
+        TORCH_CHECK(seqlens_k->scalar_type() == at::kInt,
+                    "MHA v4 seqlens_k must be int32, got ",
+                    seqlens_k->scalar_type());
+        TORCH_CHECK(seqlens_k->is_contiguous(), "MHA v4 seqlens_k must be contiguous");
+        TORCH_CHECK(seqlens_k->numel() >= shapes.batch,
+                    "MHA v4 seqlens_k needs one entry per batch: got ",
+                    seqlens_k->numel(),
+                    " for batch ",
+                    shapes.batch);
+        args.ptr_kseq.value = seqlens_k->data_ptr();
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply csrc/py_itfs_cu/asm_mha_v4_fwd.cu]: `0345bb007` + `cb9a297df`

File: aiter/ops/mha_v4.py [hunk 6]
```diff
@@ -1157,30 +1278,36 @@ def mha_v4(
                 int(v_format),
                 int(recipe.v_pack),
                 softmax_scale,
+                lse_out,
             )
+            if return_lse:
+                return out, lse_out
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply aiter/ops/mha_v4.py]: https://github.com/ROCm/aiter/pull/5798/changes/4a908c7e4d11d728689d9847bb1adfbc43f2df71

File: aiter/ops/mha_v4.py [hunk 7]
```diff
@@ -757,6 +760,30 @@ def _mha_v4_fwd_sparse_launch_fake(
...
+        (AttentionFormat.FP8_E4M3, AttentionFormat.FP6_E2M3),
+        (AttentionFormat.FP8_E4M3_FNUZ, AttentionFormat.FP6_E2M3),
+        (AttentionFormat.INT8, AttentionFormat.FP8_E4M3),
+        (AttentionFormat.INT8, AttentionFormat.FP8_E4M3_FNUZ),
+        (AttentionFormat.MXFP4, AttentionFormat.MXFP4),
+        (AttentionFormat.MXFP6, AttentionFormat.MXFP6),
+        (AttentionFormat.MXFP6, AttentionFormat.MXFP4),
+        (AttentionFormat.MXFP6, AttentionFormat.FP8_E4M3),
+        (AttentionFormat.MXFP6, AttentionFormat.FP8_E4M3_FNUZ),
+    }
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply aiter/ops/mha_v4.py]: https://github.com/ROCm/aiter/pull/5798/changes/15b591fb631a166dbf56ee6282a4d2f4e62d7838

**Comment by jcaraban** [comment]
> LGTM from the triton side overall. Please reduce the comments in `op_tests/op_benchmarks/triton/bench_sage.py`. They should be 2-3 lines maximum.

done, thanks @azaidy ! also addressed the copilot findings
